### PR TITLE
Update to new chroma to fix DB corruption issues

### PR DESCRIPTION
**Pull request by pseudotensor**
- [x] Avoid Corruption: https://github.com/chroma-core/chroma/issues/1134
- [x] Handle migration better than chroma team, allow old dbs and new dbs in same env
- [x] Improve UI-UX w.r.t. which collection being chatted with
- [x] https://github.com/chroma-core/chroma/issues/1136

Running full suite of tests.  Migration required for old databases: https://docs.trychroma.com/migration

### CODE REVIEW

File: tests/test_client_calls.py [hunk 1]
```diff
@@ -470,6 +470,7 @@ def test_client_chat_stream_langchain_steps(max_new_tokens, top_k_docs):
 
     res_dict, client = run_client(client, prompt, args, kwargs)
     assert ('a large language model' in res_dict['response'] or
+            '2oGPT is an open-source, Apache V2 project' in res_dict['response'] or
```
Comment (achraf-mer) [inline tests/test_client_calls.py:473]: nit: missing `H`

File: src/gpt_langchain.py [hunk 2]
```diff
@@ -2496,10 +2518,27 @@ def get_existing_db(db, persist_directory,
...
+                    migrate_db=True,
                     verbose=False, check_embedding=True, migrate_meta=True,
                     n_jobs=-1):
-    if load_db_if_exists and db_type == 'chroma' and os.path.isdir(persist_directory) and os.path.isdir(
-            os.path.join(persist_directory, 'index')):
+    if load_db_if_exists and db_type == 'chroma' and os.path.isdir(persist_directory):
+        if os.path.isfile(os.path.join(persist_directory, 'chroma.sqlite3')):
+            must_migrate = False
+        elif os.path.isdir(os.path.join(persist_directory, 'index')):
+            must_migrate = True
```
Comment (achraf-mer) [inline src/gpt_langchain.py:2528]: is this check enough to determine was using duckdb?

File: src/gpt_langchain.py [hunk 3]
```diff
@@ -2496,10 +2518,27 @@ def get_existing_db(db, persist_directory,
...
+        if must_migrate:
+            if migrate_db:
+                print("Detected chromadb<0.4 database, require migration, doing now....", flush=True)
+                from chroma_migrate.import_duckdb import migrate_from_duckdb
+                import chromadb
+                api = chromadb.PersistentClient(path=persist_directory)
+                did_migration = migrate_from_duckdb(api, persist_directory)
+                assert did_migration, "Failed to migrate chroma collection at %s, see https://docs.trychroma.com/migration for CLI tool" % persist_directory
+            else:
+                print("Detected chromadb<0.4 database, require migration, but , doing now....", flush=True)
```
Comment (achraf-mer) [inline src/gpt_langchain.py:2540]: maybe need to reformat the error message and hard fail.

**Review by achraf-mer — APPROVED** [review]

**Comment by pseudotensor** [comment]
```
=========================== short test summary info ============================
FAILED tests/test_client_calls.py::test_fastsys[True-None-instruct_vicuna] - ...
FAILED tests/test_client_calls.py::test_fastsys[True-None-one_shot] - assert ...
FAILED tests/test_client_calls.py::test_fastsys[True-8-instruct_vicuna] - Ass...
FAILED tests/test_client_calls.py::test_fastsys[True-8-one_shot] - AssertionE...
FAILED tests/test_client_calls.py::test_fastsys[True-4-instruct_vicuna] - ass...
FAILED tests/test_client_calls.py::test_fastsys[True-4-one_shot] - assert ('s...
FAILED tests/test_client_calls.py::test_fastsys[False-None-instruct_vicuna]
FAILED tests/test_client_calls.py::test_fastsys[False-None-one_shot] - assert...
FAILED tests/test_client_calls.py::test_fastsys[False-8-instruct_vicuna] - As...
FAILED tests/test_client_calls.py::test_fastsys[False-8-one_shot] - Assertion...
FAILED tests/test_client_calls.py::test_fastsys[False-4-instruct_vicuna] - as...
FAILED tests/test_client_calls.py::test_fastsys[False-4-one_shot] - assert ('...
FAILED tests/test_eval.py::test_eval1[False-4-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-8-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-16-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-32-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_inference_servers.py::test_openai_inference_server[openai_azure_chat-False]
FAILED tests/test_inference_servers.py::test_openai_inference_server[openai_azure_chat-True]
FAILED tests/test_long_context.py::test_gradio_long_context_uuid_key_value_retrieval[True-0.2-6000-{'type':'dynamic', 'factor':2}-h2oai/h2ogpt-4096-llama2-13b-chat]
FAILED tests/test_manual_test.py::test_chat_context - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_upload_one_file - NotImplementedError:...
FAILED tests/test_manual_test.py::test_upload_multiple_file - NotImplementedE...
FAILED tests/test_manual_test.py::test_upload_url - NotImplementedError: MANU...
FAILED tests/test_manual_test.py::test_upload_arxiv - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_upload_pasted_text - NotImplementedErr...
FAILED tests/test_manual_test.py::test_no_db_dirs - NotImplementedError: MANU...
FAILED tests/test_manual_test.py::test_upload_unsupported_file - NotImplement...
FAILED tests/test_manual_test.py::test_upload_to_UserData_and_MyData - NotImp...
FAILED tests/test_manual_test.py::test_chat_control - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_subset_only - NotImplementedError: MAN...
FAILED tests/test_manual_test.py::test_add_new_doc - NotImplementedError: MAN...
FAILED tests/test_manual_test.py::test_model_lock - NotImplementedError: MANU...
FAILED tests/test_metrics.py::test_bleurt - ImportError: To be able to use ev...
FAILED tests/test_metrics.py::test_bertscore - ImportError: To be able to use...
FAILED tests/test_metrics.py::test_wiki_split - ImportError: To be able to us...
FAILED tests/test_pipeline.py::test_pipeline1 - AssertionError: assert 'Drink...
FAILED tests/test_requirements.py::test_requirements - ValueError: ['numpy==1...
= 37 failed, 1318 passed, 278 skipped, 6 xfailed, 1 xpassed, 32 warnings in 20923.53s (5:48:43) =
```

Failures are flakiness or known issues that are not critial or are related to new features still WIP like azure openai.

File: src/gpt_langchain.py (hunk 2 above)
Comment (pseudotensor) [reply src/gpt_langchain.py:2528]: It seems so.  I had always relied upon 'index' folder before in other places.

File: tests/test_client_calls.py (hunk 1 above)
Comment (pseudotensor) [reply tests/test_client_calls.py:473]: Ya, I made test without and it passes, and I didn't want to rerun.  I'll leave for now.

**Comment by pseudotensor** [comment]
```
=========================== short test summary info ============================
FAILED tests/test_client_calls.py::test_fastsys[True-None-instruct_vicuna] - ...
FAILED tests/test_client_calls.py::test_fastsys[True-None-one_shot] - assert ...
FAILED tests/test_client_calls.py::test_fastsys[True-8-instruct_vicuna] - Ass...
FAILED tests/test_client_calls.py::test_fastsys[True-8-one_shot] - AssertionE...
FAILED tests/test_client_calls.py::test_fastsys[True-4-instruct_vicuna] - ass...
FAILED tests/test_client_calls.py::test_fastsys[True-4-one_shot] - assert ('s...
FAILED tests/test_client_calls.py::test_fastsys[False-None-instruct_vicuna]
FAILED tests/test_client_calls.py::test_fastsys[False-None-one_shot] - assert...
FAILED tests/test_client_calls.py::test_fastsys[False-8-instruct_vicuna] - As...
FAILED tests/test_client_calls.py::test_fastsys[False-8-one_shot] - Assertion...
FAILED tests/test_client_calls.py::test_fastsys[False-4-instruct_vicuna] - as...
FAILED tests/test_client_calls.py::test_fastsys[False-4-one_shot] - assert ('...
FAILED tests/test_eval.py::test_eval1[False-4-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-8-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-16-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_eval.py::test_eval1[False-32-h2oai/h2ogpt-oig-oasst1-512-6_9b]
FAILED tests/test_inference_servers.py::test_openai_inference_server[openai_azure_chat-False]
FAILED tests/test_inference_servers.py::test_openai_inference_server[openai_azure_chat-True]
FAILED tests/test_langchain_simple.py::test_langchain_simple_h2ogpt - assert ...
FAILED tests/test_langchain_units.py::test_url_more_subunit - assert 0 > 0
FAILED tests/test_long_context.py::test_gradio_long_context_uuid_key_value_retrieval[False-0.2-6000-{'type':'dynamic', 'factor':2}-h2oai/h2ogpt-4096-llama2-13b-chat]
FAILED tests/test_long_context.py::test_gradio_long_context_uuid_key_value_retrieval[True-0.2-6000-{'type':'dynamic', 'factor':2}-h2oai/h2ogpt-4096-llama2-13b-chat]
FAILED tests/test_manual_test.py::test_chat_context - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_upload_one_file - NotImplementedError:...
FAILED tests/test_manual_test.py::test_upload_multiple_file - NotImplementedE...
FAILED tests/test_manual_test.py::test_upload_url - NotImplementedError: MANU...
FAILED tests/test_manual_test.py::test_upload_arxiv - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_upload_pasted_text - NotImplementedErr...
FAILED tests/test_manual_test.py::test_no_db_dirs - NotImplementedError: MANU...
FAILED tests/test_manual_test.py::test_upload_unsupported_file - NotImplement...
FAILED tests/test_manual_test.py::test_upload_to_UserData_and_MyData - NotImp...
FAILED tests/test_manual_test.py::test_chat_control - NotImplementedError: MA...
FAILED tests/test_manual_test.py::test_subset_only - NotImplementedError: MAN...
FAILED tests/test_manual_test.py::test_add_new_doc - NotImplementedError: MAN...
FAILED tests/test_manual_test.py::test_model_lock - NotImplementedError: MANU...
= 35 failed, 1324 passed, 278 skipped, 6 xfailed, 1 xpassed, 30 warnings in 22369.24s (6:12:49) =
```

flaky otherwise

**Pull request merged**